File: #1 Convex Hull/Codes/BruteForce.cpp
#include <cstdio>
#include <algorithm>
#include <vector>
#include <cmath>
#include <limits>

class Point{
public:
	Point(long long x, long long y, int i = 0) : _x(x), _y(y), _i(i) {}
	long long x(void) const { return _x; }
	long long y(void) const { return _y; }
	int index(void) const {return _i; }
	bool operator<(const Point &q) const { return (_x == q.x()) ? _y < q.y() : _x < q.x(); }
	bool operator==(const Point &q) const { return _x == q.x() && _y == q.y() && _i == q.index(); }
private:
	long long _x;
	long long _y;
	int _i;
};

class Line{
public:
	Line(Point &p, Point &q) : _p(p), _q(q) { build(); }
	long long value(Point &p) const { return _a * p.x() + _b * p.y() + _c; }
	double distance(Point &p) const { return static_cast<double>(std::abs(value(p))) / std::sqrt(_a * _a + _b * _b); }
private:
	void build(void) { _a = _q.y() - _p.y(); _b = _p.x() - _q.x(); _c = -(_a * _p.x() + _b * _p.y()); }
	Point &_p;
	Point &_q;
	long long _a;
	long long _b;
	long long _c;
};

// main operation
void convex_hull(std::vector<Point> &points);

int main(int argc, char const *argv[]) {
	int rows;
	while(scanf("%d", &rows) != EOF) {
		std::vector<Point> points;
		points.reserve(rows * sizeof(points));
		for(int i = 1; i <= rows; ++i) {
			long long x, y;
			scanf("%I64d %I64d", &x, &y);
			points.push_back(Point(x, y, i));
		}
		convex_hull(points);
	}
	return 0;
}

void convex_hull(std::vector<Point> &points) {
	std::sort(points.begin(), points.end());
	std::vector<Point> ret;

	while (true) {
		Line L(points[0], points[1]);
		int pos = 0;
		int neg = 0;
		for (auto &point : points) {
			if(L.value(point) > 0)
				pos += 1;
			else if(L.value(point) < 0)
				neg += 1;
		}
		if(pos == 0 || neg == 0) {
			ret.push_back(points[0]);
			ret.push_back(points[1]);
		}
		std::prev_permutation(points.begin() + 2, points.end());
		if (!std::next_permutation(points.begin(), points.end()))
			break;
	}
	
	std::sort(ret.begin(), ret.end(), [&](Point p, Point q) { return p.index() < q.index(); });
	ret.erase(std::unique(ret.begin(), ret.end()), ret.end());
	printf("%d\n", ret.size());
	for(auto &i : ret)
		printf("%d\n", i.index());
}

File: #1 Convex Hull/Codes/BruteForce.py
"""
@ CSU0018 Computer Algorithms
@ Programming Assignment 01
@ Convex Hull - Brute Force
"""
import sys
SAMPLE_NO = 1 #sys.argv[1]

# Utilities
class Point:
	def __init__(self, x, y, i):
		self._x = x
		self._y = y
		self._i = i

	def __getitem__(self, coord):
		if coord == "x":
			return self._x
		elif coord == "y":
			return self._y
		elif coord == "i":
			return self._i
		else:
			assert()

class Line:
	def __init__(self, p, q):
		self._a = q["y"] - p["y"]
		self._b = p["x"] - q["x"]
		self._c = -(self._a * p["x"] + self._b * p["y"])
		self._p = p
		self._q = q

	def distance(self, point):
		return abs(self.value(point))/(self._a**2 + self._b**2)**0.5

	def value(self, point):
		return self._a * point["x"] + self._b * point["y"] + self._c

def list2points(lst):
	return [Point(x, y, i) for (i, (x, y)) in enumerate(lst, 1)]

def convex_hull_brute_force(points):
	points = list2points(points)
	vertices = []
	import itertools
	for (point1, point2) in itertools.combinations(points, 2):
		if point1['x'] == point2['x'] and point1['y'] == point2['y']:
			continue
		L = Line(point1, point2)
		pos, neg = 0, 0
		for point in points:
			if L.value(point) > 0:
				pos += 1
			elif L.value(point) < 0:
				neg += 1
		if pos == 0 or neg == 0:
			vertices += [point1, point2]
	vertices = list(set(vertices))
	vertices.sort(key = lambda p : p['i'])
	return vertices

if __name__ == '__main__':
	f_input = open('sample' + str(SAMPLE_NO) + '.txt', 'r')
	f_input.readline()
	input_data = []
	for line in f_input:
		line = line.strip('\n')
		x, y = line.split()
		input_data.append((int(x), int(y)))
	
	from time import clock
	start = clock()

	ret = convex_hull_brute_force(input_data)

	finish = clock()
	print((finish - start))
	
	f_output = open('sample' + str(SAMPLE_NO) + '_BFout.txt', 'w')
	f_output.write(str(len(ret)) + '\n')
	for p in ret:
		f_output.writelines(str(p['i']) + '\n')
#	print(len(ret))
#	for p in ret:
#		print(p['i'])